platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/managed/ManagedObjectRegistry.java

```diff
@@ -0,0 +1,220 @@
+            if (method.isAnnotationPresent(ManagedObjectCreator.class) ||
+                doesParentInterfaceMethodHaveAnnotation(instance.getClass(), method, ManagedObjectCreator.class)
+            ) {
+                if (method.getReturnType() == Void.class) {
+                    throw new IllegalArgumentException("Method " + method + " annotated with @ManagedObjectCreator must return a value");
+                }
+
+                MethodHandle handle;
+                try {
+                     handle = lookup.unreflect(method);
+                } catch (IllegalAccessException e) {
+                    throw UncheckedException.throwAsUncheckedException(e);
+                }
+
+                registerFactory(handle, instance);
+            }
+        }
+    }
+
+    public static boolean doesParentInterfaceMethodHaveAnnotation(
+        Class<?> clazz,
+        Method targetMethod,
+        Class<? extends Annotation> annotation
+    ) {
+        for (Class<?> parent : clazz.getInterfaces()) {
+            Method interfaceMethod = null;
+            try {
+                interfaceMethod = parent.getMethod(targetMethod.getName(), targetMethod.getParameterTypes());
+            } catch (NoSuchMethodException e) {
+                // Method not found in this interface, continue checking others
+            }
+            if (interfaceMethod != null && interfaceMethod.isAnnotationPresent(annotation)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private void registerFactory(MethodHandle handle, Object instance) {
+        validateMethod(handle);
+
+        MethodType type = handle.type();
+        Class<?> returnType = type.returnType();
+        FactoryRegistration existing = factories.put(returnType, new FactoryRegistration(instance, handle));
+
+        if (existing != null) {
+            throw new IllegalArgumentException("Method " + handle + " annotated with @ManagedObjectCreator conflicts with existing factory for type " + returnType);
+        }
+    }
+
+    private static void validateMethod(MethodHandle handle) {
+        MethodType type = handle.type();
+
+        // The 0th parameter is the `this` receiver parameter.
+        int actualParameterCount = type.parameterCount() - 1;
+
+        if (actualParameterCount > 2) {
+            // We only support max 2 arg factories.
+            throw new IllegalArgumentException("Method " + handle + " annotated with @ManagedObjectCreator has too many parameters");
+        }
+
+        for (int i = 0; i < actualParameterCount; i++) {
+            Class<?> parameterType = type.parameterType(i + 1);
+            if (parameterType != Class.class) {
+                throw new IllegalArgumentException("Method " + handle + " annotated with @ManagedObjectCreator must have parameters of type Class, but has parameter of type " + parameterType);
+            }
+        }
+    }
+
+    private static class FactoryRegistration {
```

Comment:
💅 If we're dealing with `MethodHandle`s, then we can use a bound handle instead. The `FactoryRegistration` can be replaced with the handle returned from `method.bindTo(instance)`. I didn't test the performance, though.